File: packages/js/src/core/org_client.ts
/**
 * OrgClient — Organization-level management client for Olbrain.
 *
 * Allows B2B clients (e.g. Sinch) to programmatically manage projects and agents
 * using an organization-level API key, without requiring Firebase authentication
 * or the Olbrain dashboard.
 *
 * Auth: Authorization: ApiKey org_live_<key>
 * Target: Olbrain Studio Backend (not the agent-cloud service)
 *
 * @example
 * ```ts
 * import { OrgClient } from '@olbrain/js-sdk';
 *
 * const client = new OrgClient({
 *   orgApiKey: 'org_live_...',
 *   studioUrl: 'https://your-studio-backend.run.app',
 * });
 *
 * const project = await client.createProject('My Project', 'org_123');
 * const agent = await client.createAgent('Support Bot', project.projectId);
 * const deployment = await client.deployAgent(agent.agentId);
 * console.log(deployment.status);
 * ```
 */

import {
  OrgClientConfig,
  ProjectInfo,
  AgentInfo,
  DeploymentInfo,
  CreateProjectOptions,
  CreateAgentOptions,
  UpdateAgentOptions,
  ListAgentsOptions,
} from './types';
import {
  OlbrainError,
  AuthenticationError,
  RateLimitError,
  NetworkError,
  ValidationError,
} from './exceptions';

const DEFAULT_TIMEOUT_MS = 30000;

/**
 * OrgClient for managing Olbrain projects and agents via an org-level API key.
 */
export class OrgClient {
  private orgApiKey: string;
  private studioUrl: string;

  constructor(config: OrgClientConfig) {
    if (!config.orgApiKey || !config.orgApiKey.startsWith('org')) {
      throw new ValidationError(
        "Invalid org API key — must start with 'org_live_' or 'org_'"
      );
    }
    if (!config.studioUrl) {
      throw new ValidationError('studioUrl is required in OrgClientConfig');
    }

    this.orgApiKey = config.orgApiKey;
    this.studioUrl = config.studioUrl.replace(/\/$/, '');
  }

  // ---------------------------------------------------------------------------
  // Project Management
  // ---------------------------------------------------------------------------

  /**
   * Create a new project inside the organization.
   */
  async createProject(
    name: string,
    organizationId: string,
    options?: CreateProjectOptions
  ): Promise<ProjectInfo> {
    if (!name) throw new ValidationError('Project name is required');

    const data = await this._request('POST', '/api/projects', {
      organization_id: organizationId,
      project_info: {
        name,
        description: options?.description ?? '',
      },
      team_access: { visibility: 'organization' },
    });

    return this._parseProject(data);
  }

  /**
   * List all projects in the organization.
   */
  async listProjects(organizationId: string): Promise<ProjectInfo[]> {
    const data = await this._request('GET', '/api/projects', undefined, {
      organization_id: organizationId,
    });
    const projects: any[] = data?.data ?? (Array.isArray(data) ? data : []);
    return projects.map((p) => this._parseProject(p));
  }

  /**
   * Delete a project by ID.
   */
  async deleteProject(projectId: string): Promise<Record<string, any>> {
    return this._request('DELETE', `/api/projects/${projectId}`);
  }

  // ---------------------------------------------------------------------------
  // Agent Management
  // ---------------------------------------------------------------------------

  /**
   * Create a new agent inside a project.
   */
  async createAgent(
    name: string,
    projectId: string,
    options?: CreateAgentOptions
  ): Promise<AgentInfo> {
    if (!name) throw new ValidationError('Agent name is required');
    if (!projectId) throw new ValidationError('projectId is required');

    const payload: Record<string, any> = {
      name,
      project_id: projectId,
      description: options?.description ?? '',
    };
    if (options?.agentType) payload.agent_type = options.agentType;
    if (options?.industry) payload.industry = options.industry;
    if (options?.useCase) payload.use_case = options.useCase;

    const data = await this._request('POST', '/api/agents', payload);
    return this._parseAgent(data);
  }

  /**
   * List agents, optionally filtered by project or organization.
   */
  async listAgents(options?: ListAgentsOptions): Promise<AgentInfo[]> {
    const params: Record<string, string> = {
      lifecycle_state: options?.lifecycleState ?? 'active',
    };
    if (options?.projectId) params.project_id = options.projectId;
    if (options?.organizationId) params.organization_id = options.organizationId;

    const data = await this._request('GET', '/api/agents', undefined, params);
    const agents: any[] = data?.agents ?? data?.data ?? (Array.isArray(data) ? data : []);
    return agents.map((a) => this._parseAgent(a));
  }

  /**
   * Get a single agent by ID.
   */
  async getAgent(agentId: string): Promise<AgentInfo> {
    const data = await this._request('GET', `/api/agents/${agentId}`);
    return this._parseAgent(data);
  }

  /**
   * Update an agent's metadata.
   */
  async updateAgent(
    agentId: string,
    options: UpdateAgentOptions
  ): Promise<Record<string, any>> {
    if (!options || Object.keys(options).length === 0) {
      throw new ValidationError('At least one field must be provided to update');
    }

    const payload: Record<string, any> = {};
    if (options.name !== undefined) payload.name = options.name;
    if (options.description !== undefined) payload.description = options.description;
    if (options.agentType !== undefined) payload.agent_type = options.agentType;
    if (options.industry !== undefined) payload.industry = options.industry;
    if (options.useCase !== undefined) payload.use_case = options.useCase;

    return this._request('PUT', `/api/agents/${agentId}`, payload);
  }

  /**
   * Delete an agent by ID.
   */
  async deleteAgent(agentId: string): Promise<Record<string, any>> {
    return this._request('DELETE', `/api/agents/${agentId}`);
  }

  /**
   * Deploy an agent (publishes its configuration to the agent-cloud).
   */
  async deployAgent(agentId: string): Promise<DeploymentInfo> {
    const data = await this._request('POST', `/api/agents/${agentId}/deploy`);
    return this._parseDeployment(data);
  }

  /**
   * Get deployment status for an agent.
   */
  async getDeploymentStatus(agentId: string): Promise<Record<string, any>> {
    return this._request('GET', `/api/agents/${agentId}/deployment-status`);
  }

  // ---------------------------------------------------------------------------
  // Response parsers
  // ---------------------------------------------------------------------------

  private _parseProject(data: any): ProjectInfo {
    const projectInfo = data?.project_info ?? {};
    return {
      projectId: data?.project_id ?? data?.id ?? '',
      name: projectInfo?.name ?? data?.name ?? '',
      organizationId: data?.organization_id ?? '',
      description: projectInfo?.description ?? data?.description ?? '',
      createdAt: data?.created_at ? String(data.created_at) : undefined,
      updatedAt: data?.updated_at ? String(data.updated_at) : undefined,
      createdBy: data?.created_by,
    };
  }

  private _parseAgent(data: any): AgentInfo {
    const basicInfo = data?.basic_info ?? {};
    const status = data?.status ?? {};
    return {
      agentId: data?.id ?? data?.agent_id ?? '',
      name: basicInfo?.name ?? data?.name ?? '',
      projectId: data?.project_id ?? '',
      organizationId: data?.organization_id ?? '',
      description: basicInfo?.description ?? data?.description ?? '',
      developmentStage: status?.development_stage ?? 'draft',
      version: data?.version ?? 1,
      createdAt: data?.created_at ? String(data.created_at) : undefined,
      updatedAt: data?.updated_at ? String(data.updated_at) : undefined,
    };
  }

  private _parseDeployment(data: any): DeploymentInfo {
    return {
      deploymentId: data?.deployment_id ?? data?.id ?? '',
      agentId: data?.agent_id ?? '',
      version: data?.version ?? 1,
      status: data?.status ?? '',
      isActive: data?.is_active ?? false,
      configFilePath: data?.config_file_path ?? '',
      createdAt: data?.created_at ? String(data.created_at) : undefined,
    };
  }

  // ---------------------------------------------------------------------------
  // Internal HTTP helper
  // ---------------------------------------------------------------------------

  private async _request(
    method: string,
    path: string,
    payload?: Record<string, any>,
    params?: Record<string, string>,
    timeout: number = DEFAULT_TIMEOUT_MS
  ): Promise<any> {
    let url = `${this.studioUrl}${path}`;

    if (params && Object.keys(params).length > 0) {
      const qs = new URLSearchParams(params).toString();
      url = `${url}?${qs}`;
    }

    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
      'Authorization': `ApiKey ${this.orgApiKey}`,
    };

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), timeout);

    try {
      const response = await fetch(url, {
        method,
        headers,
        body: payload ? JSON.stringify(payload) : undefined,
        signal: controller.signal,
      });

      clearTimeout(timeoutId);

      let responseData: any = {};
      try {
        responseData = await response.json();
      } catch {
        // non-JSON body (e.g. 204 No Content)
      }

      if (!response.ok) {
        const detail = responseData?.detail ?? responseData?.message ?? response.statusText;

        if (response.status === 401 || response.status === 403) {
          throw new AuthenticationError(detail);
        }
        if (response.status === 422) {
          throw new ValidationError(`Validation error: ${detail}`);
        }
        if (response.status === 429) {
          const retryAfter = response.headers.get('Retry-After');
          throw new RateLimitError(detail, retryAfter ? parseInt(retryAfter) : undefined);
        }
        throw new NetworkError(detail, response.status);
      }

      return responseData;
    } catch (error) {
      clearTimeout(timeoutId);

      if (error instanceof OlbrainError) throw error;

      if (error instanceof TypeError && error.message.includes('aborted')) {
        throw new OlbrainError(`Request to ${url} timed out after ${timeout}ms`);
      }

      throw new NetworkError(String(error));
    }
  }
}

File: packages/python/olbrain/org_client.py
"""
Olbrain OrgClient — Organization-level management client.

Allows B2B clients (e.g. Sinch) to programmatically manage projects and agents
using an organization-level API key, without requiring Firebase authentication
or the Olbrain dashboard.

Auth: Authorization: ApiKey org_live_<key>
Target: Olbrain Studio Backend (not the agent-cloud service)

Example:
    >>> from olbrain import OrgClient
    >>>
    >>> client = OrgClient(
    ...     org_api_key="org_live_...",
    ...     studio_url="https://your-studio-backend.run.app"
    ... )
    >>>
    >>> project = client.create_project("My Project", organization_id="org_123")
    >>> agent = client.create_agent("Support Bot", project_id=project.project_id)
    >>> deployment = client.deploy_agent(agent.agent_id)
    >>> print(deployment.status)
"""

import logging
import os
import requests
from typing import Dict, Any, List, Optional

from .session import ProjectInfo, AgentInfo, DeploymentInfo
from .exceptions import (
    OlbrainError,
    AuthenticationError,
    NetworkError,
    RateLimitError,
    ValidationError
)

logger = logging.getLogger(__name__)

DEFAULT_TIMEOUT = 30


class OrgClient:
    """
    Client for managing Olbrain projects and agents via an organization API key.

    Uses Authorization: ApiKey <org_live_key> against the studio backend.
    The existing AgentClient (for chatting with agents) is separate and unchanged.

    Args:
        org_api_key: Organization API key (must start with 'org_live_' or 'org_')
        studio_url: Base URL of the Olbrain Studio Backend.
                    Falls back to OLBRAIN_STUDIO_URL env var if not provided.
    """

    def __init__(
        self,
        org_api_key: str,
        studio_url: Optional[str] = None
    ):
        valid_prefixes = ('org_live_', 'org_')
        if not org_api_key or not org_api_key.startswith(valid_prefixes):
            raise ValueError(
                "Invalid org API key — must start with 'org_live_' or 'org_'"
            )

        self.org_api_key = org_api_key
        self.studio_url = (
            studio_url
            or os.environ.get('OLBRAIN_STUDIO_URL', '')
        ).rstrip('/')

        if not self.studio_url:
            raise ValueError(
                "studio_url is required. Pass it directly or set the "
                "OLBRAIN_STUDIO_URL environment variable."
            )

        logger.info(f"OrgClient initialized against {self.studio_url}")

    # -------------------------------------------------------------------------
    # Project Management
    # -------------------------------------------------------------------------

    def create_project(
        self,
        name: str,
        organization_id: str,
        description: str = ""
    ) -> ProjectInfo:
        """
        Create a new project inside the organization.

        Args:
            name: Project name
            organization_id: Organization ID this project belongs to
            description: Optional project description

        Returns:
            ProjectInfo with the created project's details
        """
        if not name:
            raise ValidationError("Project name is required")

        payload = {
            "organization_id": organization_id,
            "project_info": {
                "name": name,
                "description": description
            },
            "team_access": {
                "visibility": "organization"
            }
        }

        data = self._request("POST", "/api/projects", payload)
        return ProjectInfo.from_dict(data)

    def list_projects(self, organization_id: str) -> List[ProjectInfo]:
        """
        List all projects in the organization.

        Args:
            organization_id: Organization ID to filter by

        Returns:
            List of ProjectInfo objects
        """
        data = self._request(
            "GET", "/api/projects",
            params={"organization_id": organization_id}
        )
        projects = data.get("data", data if isinstance(data, list) else [])
        return [ProjectInfo.from_dict(p) for p in projects]

    def delete_project(self, project_id: str) -> Dict[str, Any]:
        """
        Delete a project by ID.

        Args:
            project_id: Project ID to delete

        Returns:
            Dict with success confirmation
        """
        return self._request("DELETE", f"/api/projects/{project_id}")

    # -------------------------------------------------------------------------
    # Agent Management
    # -------------------------------------------------------------------------

    def create_agent(
        self,
        name: str,
        project_id: str,
        description: str = "",
        agent_type: str = None,
        industry: str = None,
        use_case: str = None
    ) -> AgentInfo:
        """
        Create a new agent inside a project.

        Args:
            name: Agent name
            project_id: Project ID the agent belongs to
            description: Optional agent description
            agent_type: Optional agent type
            industry: Optional industry
            use_case: Optional use case

        Returns:
            AgentInfo with the created agent's details
        """
        if not name:
            raise ValidationError("Agent name is required")
        if not project_id:
            raise ValidationError("project_id is required")

        payload: Dict[str, Any] = {
            "name": name,
            "project_id": project_id,
            "description": description
        }
        if agent_type:
            payload["agent_type"] = agent_type
        if industry:
            payload["industry"] = industry
        if use_case:
            payload["use_case"] = use_case

        data = self._request("POST", "/api/agents", payload)
        return AgentInfo.from_dict(data)

    def list_agents(
        self,
        project_id: str = None,
        organization_id: str = None,
        lifecycle_state: str = "active"
    ) -> List[AgentInfo]:
        """
        List agents, optionally filtered by project or organization.

        Args:
            project_id: Optional project filter
            organization_id: Optional organization filter
            lifecycle_state: Filter by state (default: active)

        Returns:
            List of AgentInfo objects
        """
        params: Dict[str, Any] = {"lifecycle_state": lifecycle_state}
        if project_id:
            params["project_id"] = project_id
        if organization_id:
            params["organization_id"] = organization_id

        data = self._request("GET", "/api/agents", params=params)
        agents = data.get("agents", data.get("data", []))
        return [AgentInfo.from_dict(a) for a in agents]

    def get_agent(self, agent_id: str) -> AgentInfo:
        """
        Get a single agent by ID.

        Args:
            agent_id: Agent ID

        Returns:
            AgentInfo
        """
        data = self._request("GET", f"/api/agents/{agent_id}")
        return AgentInfo.from_dict(data)

    def update_agent(
        self,
        agent_id: str,
        name: str = None,
        description: str = None,
        agent_type: str = None,
        industry: str = None,
        use_case: str = None
    ) -> Dict[str, Any]:
        """
        Update an agent's metadata.

        Args:
            agent_id: Agent ID to update
            name: New name (optional)
            description: New description (optional)
            agent_type: New type (optional)
            industry: New industry (optional)
            use_case: New use case (optional)

        Returns:
            Success response dict
        """
        payload: Dict[str, Any] = {}
        if name is not None:
            payload["name"] = name
        if description is not None:
            payload["description"] = description
        if agent_type is not None:
            payload["agent_type"] = agent_type
        if industry is not None:
            payload["industry"] = industry
        if use_case is not None:
            payload["use_case"] = use_case

        if not payload:
            raise ValidationError("At least one field must be provided to update")

        return self._request("PUT", f"/api/agents/{agent_id}", payload)

    def delete_agent(self, agent_id: str) -> Dict[str, Any]:
        """
        Delete an agent by ID.

        Args:
            agent_id: Agent ID to delete

        Returns:
            Dict with success confirmation and delete_type (hard/soft)
        """
        return self._request("DELETE", f"/api/agents/{agent_id}")

    def deploy_agent(self, agent_id: str) -> DeploymentInfo:
        """
        Deploy an agent (publishes its configuration to the agent-cloud).

        Args:
            agent_id: Agent ID to deploy

        Returns:
            DeploymentInfo with deployment_id, version, and status
        """
        data = self._request("POST", f"/api/agents/{agent_id}/deploy")
        return DeploymentInfo.from_dict(data)

    def get_deployment_status(self, agent_id: str) -> Dict[str, Any]:
        """
        Get deployment status for an agent.

        Args:
            agent_id: Agent ID

        Returns:
            Dict with hasDeployment, deploymentCount, latestDeploymentId
        """
        return self._request("GET", f"/api/agents/{agent_id}/deployment-status")

    # -------------------------------------------------------------------------
    # Context Manager
    # -------------------------------------------------------------------------

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        pass  # No persistent connections to close

    def __repr__(self) -> str:
        return f"OrgClient(studio_url={self.studio_url})"

    # -------------------------------------------------------------------------
    # Internal HTTP helper
    # -------------------------------------------------------------------------

    def _request(
        self,
        method: str,
        path: str,
        payload: Dict[str, Any] = None,
        params: Dict[str, Any] = None,
        timeout: int = DEFAULT_TIMEOUT
    ) -> Any:
        """
        Make an authenticated HTTP request to the studio backend.

        Auth: Authorization: ApiKey <org_api_key>
        """
        url = f"{self.studio_url}{path}"
        headers = {
            "Authorization": f"ApiKey {self.org_api_key}",
            "Content-Type": "application/json"
        }

        try:
            response = requests.request(
                method=method,
                url=url,
                headers=headers,
                json=payload,
                params=params,
                timeout=timeout
            )
            self._handle_errors(response, f"{method} {path}")
            return response.json()

        except requests.exceptions.ConnectionError as e:
            raise NetworkError(f"Cannot connect to studio backend at {self.studio_url}: {e}")
        except requests.exceptions.Timeout:
            raise NetworkError(f"Request to {url} timed out after {timeout}s")
        except requests.exceptions.RequestException as e:
            raise NetworkError(f"Network error: {e}")

    def _handle_errors(self, response: requests.Response, operation: str):
        """Raise typed exceptions for HTTP error responses."""
        if response.status_code < 400:
            return

        try:
            detail = response.json().get("detail", response.text)
        except Exception:
            detail = response.text

        if response.status_code == 401:
            raise AuthenticationError(f"Invalid or expired org API key")
        if response.status_code == 403:
            raise AuthenticationError(f"Permission denied: {detail}")
        if response.status_code == 404:
            raise OlbrainError(f"Not found: {detail}")
        if response.status_code == 422:
            raise ValidationError(f"Validation error: {detail}")
        if response.status_code == 429:
            raise RateLimitError(f"Rate limit exceeded: {detail}")
        raise OlbrainError(
            f"Request failed [{response.status_code}] during {operation}: {detail}"
        )
